### PR TITLE
Check all features and targets in CI when linting

### LINKED ISSUE #1644: We've exceeded the concurrency limit for GitHub Actions
We are limited to [20 concurrent jobs](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration), meaning that any additional jobs will be queued until a runner is available. This will obviously drag out the total runtime of the workflow, and we should consider restructuring things.

### LINKED ISSUE #1693: Add an example and/or other CI check to verify `defmt` and other features

### PR DESCRIPTION
**Pull request by MabezDev**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [ ] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/CONTRIBUTING.md) and followed its instructions.

#### Description

What this doesn't solve just yet:

* log/defmt, checking, I've gone with defmt for now, but ideally I'd like to check both but this would require double the CI time.
* ~all the packages, this _mostly_ tackles esp-hal and esp-wifi, but I'd like 100% coverage~ I think this covers everything now, at the very least each package is built for each target.

Closes #1644
Closes #1693

### CODE REVIEW

**Comment by MabezDev** [comment]
Reminder to myself: this PR has some breaking changes (enum renaming mostly) that needs a changelog entry

File: esp-wifi/src/wifi/mod.rs
Comment (MabezDev) [inline esp-wifi/src/wifi/mod.rs:2702]: This was a bit too much work for this PR, we should file another issue for these.

**Comment by MabezDev** [comment]
This makes the clippy job take 16minutes, but I don't really see a way to reduce this. If we want everything checked, then it takes time.

**Comment by bjoernQ** [comment]
> This makes the clippy job take 16minutes, but I don't really see a way to reduce this. If we want everything checked, then it takes time.

I think there is no way around that

File: esp-wifi/src/wifi/mod.rs
Comment (MabezDev) [reply esp-wifi/src/wifi/mod.rs:2702]: Jk, wasn't too hard

**Comment by MabezDev** [comment]
On second review, I don't _think_ this PR has any breaking changes, but please check my work :D

**Review by bjoernQ — APPROVED** [review]
Thanks!

**Comment by bjoernQ** [comment]
Haven't seen anything looking sus

There is one warning:
```
warning: constant `ESP32P4_TOML` is never used
  --> esp-metadata\src\lib.rs:10:7
   |
10 | const ESP32P4_TOML: &str = include_str!("../devices/esp32p4.toml");
   |       ^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default
```

File: esp-hal/src/uart.rs
Comment (JurajSadel) [inline esp-hal/src/uart.rs:1858]:
```suggestion
```

**Review by JurajSadel — APPROVED** [review]
LGTM, left a nitpick suggestion, feel free to ignore. Thanks!

File: esp-hal/src/uart.rs
Comment (MabezDev) [reply esp-hal/src/uart.rs:1858]: What's the suggested change? I just see an empty line? :D

Comment (JurajSadel) [reply esp-hal/src/uart.rs:1858]: Haha, that's the suggested change - remove the empty line :D

Comment (MabezDev) [reply esp-hal/src/uart.rs:1858]: rustfmt put it there, so I'll leave it be :D

**Comment by MabezDev** [comment]
I think I might be able to trim some old CI jobs now as they should be covered by clippy.

And provided we are still building the examples we should still detect linker issues.

Let me try.

**Comment by MabezDev** [comment]
Hmm, that worked. Now I'm wondering if I can split the clippy job into the esp-hal matrix, which _might_ make the clippy job quicker as most of the crate should be built? Let's see :D

**Comment by MabezDev** [comment]
It helped a little bit I think, it shaved a few minutes of some of the jobs at least.

This is now ready for a final review. Before merging we need to remove the required checks on the jobs that no longer exist, and subsequent PRs should be rebased.

File: .github/workflows/ci.yml
Comment (SergioGasquez) [inline .github/workflows/ci.yml:169]:
```suggestion
      - name: msrv Xtensa (esp-hal)
```

**Review by SergioGasquez — APPROVED** [review]
LGTM! I like the improvements, thanks!

File: esp-metadata/Cargo.toml
Comment (jessebraham) [inline esp-metadata/Cargo.toml:12]: Why were these dependencies added?

**Pull request merged**
